### PR TITLE
Containerd image and snapshot backend integration

### LINKED ISSUE #26146: cancelling a pull causes partially or fully downloaded layers to be discarded
I'm working from an unreliable network connection and sometimes I need to pull large images. When I'm pulling a large image and the wifi disconnects and then reconnects, the download hangs. Then I have to ctrl+c to cancel the pull and try it again. When I do this, all layers pulled during the previous attempt have to be re-pulled from scratch. This means that if I never have a stable connection for long enough to pull the entire image without interruptions I can't pull the image at all.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1-rc2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   236317f-unsupported
 Built:        Thu Aug 18 17:11:17 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1-rc2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   236317f-unsupported
 Built:        Thu Aug 18 17:11:17 2016
 OS/Arch:      linux/amd64
```
This is a slightly hacked up version, but the changes are not related. It's basically 1.12.1.

**Output of `docker info`:**

```
Containers: 10
 Running: 10
 Paused: 0
 Stopped: 0
Images: 110
Server Version: 1.12.1-rc2
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.7.2-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.708 GiB
Name: compooter
ID: GKSI:YHZC:YPNI:PAGO:EYPD:AYME:K6BY:BAIH:EBKQ:SOAK:WUFP:6MJM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 62
 Goroutines: 113
 System Time: 2016-08-30T08:12:15.46410461-07:00
 EventsListeners: 1
Username: viktorstanchev
Registry: https://index.docker.io/v1/
Cluster Store: etcd://172.17.0.1:12379
Cluster Advertise: 172.17.0.1:12376
Insecure Registries:
 192.168.0.106
 dtr2-2061409770.us-west-2.elb.amazonaws.com
 new-1428125395.us-west-2.elb.amazonaws.com
 172.17.0.1
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical laptop

**Steps to reproduce the issue:**
1. pull a large image
2. disconnect from the network and re-connect
3. all layer downloads hang; ctrl+c to cancel
4. pull again

**Describe the results you received:**
pulls from scratch

**Describe the results you expected:**
continues where it left off

**Additional information you deem important (e.g. issue happens only occasionally):**
This behaviour makes sense because otherwise a mechanism for cleaning up unused partial downloads is necessary, but I think that keeping these partial downloads around for about a day or some other arbitrary time period is reasonable.

### LINKED ISSUE #38043: Proposal: containerd image integration
## Summary
Replace Moby image backend with containerd. This will move ownership of image related code outside of Moby allowing for a large code and scope reduction in Moby. Many refactorings can be done before doing the switch, but changing the backend can only be done in large increments due to the metadata storage architecture difference in containerd. 

### Overview
- Image storage moved to containerd metadata database
- Migration of existing metadata
- All image requests use containerd client
- Distribution code removed and replaced with containerd client
- Containerd’s snapshotters supported
- Graph drivers scoped down (no more mounting/diffing)
- Graph drivers supported as containerd snapshotter plugins

## Plan
This change is deep, in some cases we can completely bypass existing packages and use containerd (i.e. distribution), while in others we will need to write new code to integrate with existing Docker interfaces. There will likely be a few large PRs since Docker will not be able to be in a functional state unless the data storage is consistently used by all commands. This plan is an attempt to break down the work in individual PRs which can be merged in isolation while preserving as much functionality as possible. Some PRs may temporarily break upgrades since we will want to begin testing and integrating with some of these changes before the upgrade migration code is complete.

I have broken this down into phases to better understand the relationship between work items. There is no definition for the end of a phase but rather the work in each phase should be able to be completed concurrently.

### Phase 1
- libcontainerd simplification and windows runtime integration
- daemon updates
  - replace plugin executor interface with containerd client
  - replace image backend and distribution with containerd client
    - includes all image operations: ls, rm, pull, push, etc
    - unpack from content store into layer store
- minimizing graph driver scope
  - move mount to layer store and utilize containerd mount package
  - move diff to layer store and utilize containerd diff packages
### Phase 2
- move libcontainerd into daemon
- image metadata migration
  - support migrating existing docker metadata into containerd metadata store
- graph drivers updated to snapshotter interface
  - layer store communicating with containerd snapshotter
### Phase 3
- replace layer store with snapshotter
  - graph drivers exposed as containerd snapshotter plugins
  - support layer recreation into content store from tar split on push
- update storage driver configuration to reflect snapshotter preference

## Related PRs

 - [x] ~https://github.com/moby/moby/pull/37932~ https://github.com/moby/moby/pull/38168 Update containerd to 1.2x
 - [x] https://github.com/moby/moby/pull/37874 Remove the rest of v1 manifest support
   - [x] https://github.com/moby/moby/pull/39365 Keep but deprecate registry v2 schema1 logic and revert to libtrust-key-based engine ID
   - [ ] https://github.com/moby/moby/pull/39384 Remove v2 schema1 push
   - [ ] https://github.com/moby/moby/pull/39371 Remove more registry v1 code
 - [ ] https://github.com/moby/moby/pull/38738 Containerd image and snapshot backend integration

### Related project

https://github.com/dmcgowan/docker/projects/1 containerd-integration

### PR DESCRIPTION
**Pull request by dmcgowan**
Opening this up as a draft as there are many things left to do before this is mergeable.

Highlights
- Image and reference store deprecated for containerd backend
- ImageID can reference manifests, manifest lists, or image configs (allowing better support for multi-arch images)
- Support for multiple unpacked storage backends (images point to their backend)
- Multi-arch Image support

Major things left to do
- Migration of existing image store data
- Build code
- Artifact creation on push and commit
- Snapshot support as an option for images to point to
- Many, many TODOs

Implements #38043

Does not include, both snapshotters and graphdriver interfaces will be supported in this PR
* Graph drivers scoped down (no more mounting/diffing)
* Graph drivers supported as containerd snapshotter plugins

Closes #26146
Closes #38043 
Closes https://github.com/moby/moby/pull/33355
Closes https://github.com/docker/cli/pull/122

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Is this planned to be merged before Docker v19.03?

**Comment by dmcgowan** [comment]
> Is this planned to be merged before Docker v19.03?

That is desirable but unrealistic. I want to complete as many or all of the major things by that time frame, but until all the functionality is complete it is too hard to pin down an exact date for "stable enough to merge". 

It is very important we get the design parts mentioned in "highlights" correct so we aren't in a position where we need to correct storage code later (adding more support permutations and instability). Please comment on the design and help wherever possible will certainly help stabilize this change.

File: daemon/images/image.go [hunk 1]
```diff
@@ -1,11 +1,39 @@
...
 	"github.com/docker/docker/image"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+const (
+	// LabelImageID refers to the image ID used by Docker
+	// Deprecate this to support multi-arch images
+	LabelImageID = "docker.io/image.id"
```
Comment (AkihiroSuda) [inline daemon/images/image.go]: nit: `mobyproject.org`? `docker.com`?

**Comment by cpuguy83** [comment]
Pushed a rebase.

File: daemon/images/image.go (hunk 1 above)
Comment (thaJeztah) [reply daemon/images/image.go]: If these can conflict with user-provided labels; reverse-dns notation was originally marked as reserved; https://docs.docker.com/config/labels-custom-metadata/#key-format-recommendations

> The `com.docker.*`, `io.docker.*`, and `org.dockerproject.*` namespaces are reserved by Docker for internal use.

**Comment by andrewrynhard** [comment]
Is this planned for Docker 19.03? I have a use case where I would like to `--output=type=docker,dest=example.tar`, but this is not supported for multi-arch containers.

**Comment by thaJeztah** [comment]
No, this is not for 19.03; much work to do yet before this can be merged

**Comment by dt-rush** [comment]
Is there any way we might separate out the multi-arch image support and merge that sooner? This seems it's a long time running since February and has slowed down a fair bit. Lack of multi-arch image support is preventing `docker buildx --load` from loading manifest list "images" into the docker daemon.

**Comment by carlosedp** [comment]
Any news about this? It's pretty important for local development with buildx avoiding problems like https://github.com/docker/buildx/issues/59.
Thanks!

**Comment by CameronMcWilliam** [comment]
Also interested to see if this has made any progress recently! :)

**Comment by t0rr3sp3dr0** [comment]
@dmcgowan do you plan to continue developing this?

**Comment by milahu** [comment]
oh, i see. im still surprised, cos `docker load < oci.tar` would be trivial to implement (no need for containerd), since the only difference between oci and docker are the file locations, e.g. `layerhash.tar` vs `blobs/sha256/layerhash` (please tell me im terribly wrong ^^). imho oci is better, cos `sha256sum` is much faster than `tarhash`

**Comment by tianon** [comment]
I don't think you're wrong, but this is not the right place to discuss it 😅

(A new PR would be very interesting 👀)

**Comment by RoyHP** [comment]
This would be great to see, especially now with Apple Silicon creating a barrier to entry for people trying to run amd64 containers in the cloud but arm64 locally on their machines.

**Comment by dmcgowan** [comment]
I'm going to close this one since I don't want anyone to get the impression I am still working on it. The branches live on in my fork if anyone wants to play around with it. Reaching full behavior parity and just keeping this branch updated is non-trivial.

**Comment by pkit** [comment]
@tianon 
> I don't think you're wrong, but this is not the right place to discuss it sweat_smile

I'm not sure how come it's not the right place to discuss it? 
This issue is a direct descendant of the original issue #25779 (opened in 2016!) that wanted exactly that: `docker load < oci.tar`
But okay, let's reopen #25779 then.

**Pull request closed without merging**